docs/features/incremental-generators.cookbook.md

```diff
@@ -0,0 +1,656 @@
+        });
+    }
+}
+```
+
+**Alternative Solution**: If you are also providing a library to your users, in addition to a source generator, simply have that library include the
+attribute definition.
+
+### Additional file transformation
+
+**User scenario:** As a generator author I want to be able to transform an external non-C# file into an equivalent C# representation.
+
+**Solution:** Use the `AdditionalTextsProvider` to filter for and retrieve your files. Transform them into the code you care about, then
+register that code with the solution.
+
+**Example:**
+
+```csharp
+[Generator]
+public class FileTransformGenerator : IIncrementalGenerator
+{
+    public void Initialize(IncrementalGeneratorInitializationContext context)
+    {
+        var pipeline = context.AdditionalTextsProvider.Select(static (text, cancellationToken) =>
+        {
+            if (!text.Path.EndsWith("*.xml"))
+            {
+                return default;
+            }
+
+            return (Name: Path.GetFileName(text.Path), Code: MyXmlToCSharpCompiler.Compile(text.GetText(cancellationToken)));
+        })
+        .Where((pair) => pair is not ((null, _) or (_, null)));
+
+        context.RegisterSourceOutput(pipeline,
+            static (context, pair) => 
+                // Note: this AddSource is simplified. You will likely want to include the path in the name of the file to avoid
+                // issues with duplicate file names in different paths in the same project.
+                context.AddSource($"{pair.Name}generated.cs", SourceText.From(pair.Code, Encoding.UTF8)));
+    }
+}
+```
+
+### Augment user code
+
+**User scenario:** As a generator author I want to be able to inspect and augment a user's code with new functionality.
+
+**Solution:** Require the user to make the class you want to augment be a `partial class`, and mark it with a unique attribute.
+Provide that attribute in a `RegisterPostInitializationOutput` step. Register for callbacks on that attribute with
+`ForAttributeWithMetadataName` to collect the information needed to generate code, and use tuples (or create an equatable model)
+ to pass along that information. That information should be extracted from syntax and symbols; **do not put syntax or symbols into
+ your models**.
+e
+**Example:**
+
+```csharp
+public partial class UserClass
+{
+    [Generate]
+    public partial void UserMethod();
+}
+```
+
+```csharp
+[Generator]
+public class AugmentingGenerator : IIncrementalGenerator
+{
+    public void Initialize(IncrementalGeneratorInitializationContext context)
+    {
+        context.RegisterPostInitializationOutput(static postInitializationContext =>
+            postInitializationContext.AddSource("myGeneratedFile.cs", SourceText.From("""
+                using System;
+                namespace GeneratedNamespace
+                {
+                    [AttributeUsage(AttributeTargets.Method)]
+                    public class GeneratedAttribute : Attribute
```

Comment:
```suggestion
                    public sealed class GeneratedAttribute : Attribute
```

Reply:
Why did you choose `public` here vs. `internal`? Would default to `internal` unless there was a specific reason this should be accessible to calling assemblies.

Reply:
Most of these examples were copied, so I didn't make explicit decisions. That being said, I agree with making them `internal sealed`, and will go update the old cookbook as well.